### ISSUE TITLE
Fix S4049: GetEnumerator should be white-listed

### ISSUE THREAD

**Issue by BrightLight** [SonarSource/sonar-csharp#1593]
### Description

[csharpsquid:S4049](https://sonarcloud.io/organizations/brandonmarcum-github/rules?open=csharpsquid%3AS4049&rule_key=csharpsquid%3AS4049) ("Properties should be preferred") reports an issue on `GetEnumerator` even if it meets the criteria to be used in a `foreach` statement (see Related information). Converting it into a property would break this.

### Repro steps
Use this code snipped:
```csharp
  public interface IFoo
  {
    IMyEnumerator GetEnumerator();
  }

  public interface IMyEnumerator
  {
    int Current { get; }
    bool MoveNext();
  }

  public class Bar
  {
    private void SomeMethod(IFoo foo)
    {
      foreach (var x in foo)
      {
        //
      }
    }
  }
```

### Expected behavior

Don't report on `IFoo.GetEnumerator()`.

### Actual behavior

csharpsquid:S4049 creates an issue for `IFoo.GetEnumerator()`

### Known workarounds

No known workarounds except manually setting this issue to False Positive once it shows up in SonarQube.

### Related information
See [Essential .NET - Understanding C# foreach Internals and Custom Iterators with yield](https://msdn.microsoft.com/en-us/magazine/mt797654.aspx)
> foreach without IEnumerable: C# doesn’t require that IEnumerable/IEnumerable<T> be implemented to iterate over a data type using foreach. Rather, the compiler uses a concept known as duck typing; it looks for a GetEnumerator method that returns a type with a Current property and a MoveNext method. Duck typing involves searching by name rather than relying on an interface or explicit method call to the method.

**Comment by valhristov** [comment]
Hi @BrightLight, why would you prefer a custom interface with a `GetEnumerator` method over the standard `IEnumerable<T>` or `IEnumerable`?

**Comment by BrightLight** [comment]
Hi @valhristov,

Just to be clear: I'm not a fan of this. I just wanted to point out that it's a compiler detail which this rule currently doesn't consider. To me it is a form of pattern matching: if the requirements that `foreach` needs are met, allow it to be used in `foreach`. From what [I read](https://stackoverflow.com/questions/6368967/duck-typing-in-the-c-sharp-compiler) there are other scenarios where Microsoft uses this method.

Sadly, we do have code that needs it. We have a native Delphi application that uses .Net plugins.
We define the COM interfaces in .Net and generate the .pas (Delphi) files using regasm and tlibimp. For reasons we couldn't figure out this doesn't work for interfaces that "derive from" (or "demand") other interfaces. The information is simply lost. For example, `public interface IFoo : IEnumerable`
would translate to Delphi as `IFoo = interface(IUnknown)`.
Therefore we have a rule that COM-interfaces cannot "derive" from other interfaces.

A second argument is that we can't use generic types in COM. Using this "pattern matching" we can still have type safety on both .Net and Delphi side.
Dumbed-down example:

```pascal
  IPersonList = interface(IUnknown)
    ['{5B7B7694-9DE0-4339-89C1-08605EDCE0C3}']
    function Get_Count: Integer; safecall;
    function Get_Item(index: Integer): IPerson; safecall;
    procedure _Set_Item(index: Integer; const pRetVal: IPerson); safecall;
    procedure Add(const Item: IPerson); safecall;
    procedure Clear; safecall;
    function Remove(const Item: IPerson): WordBool; safecall;
    function Contains(const Item: IPerson): WordBool; safecall;
    function GetEnumerator: IPersonEnumerator; safecall;
    function First: IPerson; safecall;
    function Last: IPerson; safecall;
    property Count: Integer read Get_Count;
    property Item[index: Integer]: IPerson read Get_Item write _Set_Item; default;
  end;

  IPersonEnumerator = interface(IUnknown)
    ['{1BD266C4-95EE-41FA-9483-D2D76B91F7DE}']
    function Get_Current: IPerson; safecall;
    function MoveNext: WordBool; safecall;
    procedure Reset; safecall;
    property Current: IPerson read Get_Current;
  end;
```

**Comment by valhristov** [comment]
Hi @BrightLight,

Thanks for the detailed explanation! It makes sense and we will add the exception to the rule with the next release of SonarC#. (I would have said that in my first reply, but I was curious what could be the reason for using a nonstandard interface :))

**Comment by Evangelink** [comment]
Hi @BrightLight,

Just to be sure we understand correctly your request. You suggest to only add `GetEnumerator` as part of a white-list?

**Comment by BrightLight** [comment]
Hi @Evangelink.

First of all, "thanks" that you are willing to fix this!
I guess the approach depends on how much work you want to put into this. :) A "good enough" solution could be to simply exclude methods with the name "GetEnumerator()" (that don't accept any arguments).
The most precise solution would probably require do do the same duck-typing check as the compiler: "it looks for a GetEnumerator method that returns a type with a Current property and a MoveNext method." (quote from [Essential .NET - Understanding C# foreach Internals and Custom Iterators with yield](https://msdn.microsoft.com/en-us/magazine/mt797654.aspx))

@valhristov sure, no problem. Having to explain our code/requirements is a good opportunity to reflect on how we do certain things the way we do them :)
